### PR TITLE
[low-code connectors] Refactor paginator component so it owns the request options to set

### PR DESCRIPTION
**Pull request by girarda**
## What
* Refactor paginator so it defines the request options and paths to set if needed

The main paginator that came out of this is the LimitPaginator, which sets 2 request option (generally header, req param, or path): limit (page size) and the page token (generally offset, page number, or a cursor to follow)

`RequestOption`s have 2 fields:
- pass_by: defines where the value will be set
- field_name: the option's key
so RequestOption(request_header, "page_size") will indicate that the paginator should set a request header with key="page_size"

Sample config 1:
* limit paginator
* fetches 10 records at a time
* sets "limit" request param to 10
* gets the next_page_token from "{{ decoded_response._metadata.next }}"
* sets next_page_token to "page_size" request param
```
      paginator:
        type: "LimitPaginator"
        page_size: 10
        limit_option:
          pass_by: request_parameter
          field_name: page_size
        page_token_option:
          pass_by: path
        pagination_strategy:
          type: "CursorPagination"
          cursor_value: "{{ decoded_response._metadata.next }}"
```

Sample config 2:
* limit paginator
* fetches 5 records at a time
* sets "limit" request parameter to 5
* increments a record counter and set the request parameter "offset" to the value of the counter
```
      paginator:
        type: "LimitPaginator"
        page_size: 5
        limit_option:
          pass_by: request_parameter
          field_name: limit
        pagination_strategy:
          type: "OffsetIncrement"
        page_token_option:
          pass_by: "request_parameter"
          field_name: "offset"
```

Sample config 3
* limit paginator
* fetches 5 records at a time
* sets "limit" request parameter to 5
* increments a page counter and set the request parameter "page" to the value of the counter
```
      paginator:
        type: "LimitPaginator"
        page_size: 5
        limit_option:
          pass_by: request_parameter
          field_name: limit
        pagination_strategy:
          type: "PageIncrement"
        page_token_option:
          pass_by: "request_parameter"
          field_name: "page"
```

## How

The gist of the change is
* separate the pagination into 4 concepts
1. stop condition: when to stop paginating
2. pagination strategy: given a response, what is the next_page_token
3. where to set the next page token
4. any extra request header or parameter to set

`LimitPaginator` sets a "limit" field describing how much records to fetch and how to tell the api, and a stop condition so it stops fetching records if the last response returned fewer records than limit

* PaginationStrategy defines how to get the next_page_token
* OffsetIncrement will increment the offset by the number of records
* PageIncrement will increment the page count
* CursorPaginationStrategy will evaluate the cursor using interpolation on request and records

Sample config 4
* cursor paginator
* stops paginating when "{{ decoded_response._metadata.self == decoded_response._metadata.next }}"
* sets "page_size" request param to 10
* updates the path with "{{ decoded_response._metadata.next }}"
```
      paginator:
        type: "LimitPaginator"
        page_size: 10
        limit_option:
          pass_by: request_parameter
          field_name: limit
        page_token_option:
          option_type: path
        pagination_strategy:
          type: "CursorPagination"
          cursor_value: "{{ decoded_response._metadata.next }}"
         stop_condition: "{{ decoded_response._metadata.self == decoded_response._metadata.next }}"
```

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_option.py`
2. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/paginator.py`
3. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py`
4. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/pagination_strategy.py`
5. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/limit_paginator.py`
6. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/offset_increment.py`
7. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/page_increment.py`
8. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/cursor_pagination_strategy.py`
9. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/no_pagination.py`
10. `airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py`
11. `airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/class_types_registry.py`
12. `airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/default_implementation_registry.py`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by brianjlai — COMMENTED** [review]
Changes look good to me. I think there were a few small things and nit or two.

Also, given the complexity of the paginator now, it might be helpful to break down the folders a little more. A `paginators/strategy` folder would help organize things more.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py [hunk 1]
```diff
@@ -2,40 +2,179 @@
...
+            **self._request_options_provider.request_params(stream_state=None, stream_slice=None, next_page_token=None),
+        }
+
+    def request_headers(self) -> Mapping[str, str]:
+        return {
+            **self._get_request_options(RequestOptionType.header),
+            **self._request_options_provider.request_headers(stream_state=None, stream_slice=None, next_page_token=None),
+        }
+
+    def request_body_data(self) -> Union[Mapping[str, Any], str]:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py:103]: as per in person convo, we'll get rid of the union and just return a mapping for simplicity

**Review by brianjlai — APPROVED** [review]
looks good to me!

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py (hunk 1 above)
Comment (sherifnada) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py:103]: agreed that's a better approach

**Review by sherifnada — COMMENTED** [review]
Generally looks great, just some questions and I think we should be ready to go tomorrow from my side

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_option.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    body_data = "body_data"
+    body_json = "body_json"
+
+
+class RequestOption:
+    """
+    Describes an option to set on a request
+    """
+
+    def __init__(self, pass_by: Union[RequestOptionType, str], field_name: Optional[str] = None):
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_option.py]: yes. added a test to confirm we don't need to union with str

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/limit_paginator.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+from airbyte_cdk.sources.declarative.requesters.paginators.conditional_paginator import ConditionalPaginator
+from airbyte_cdk.sources.declarative.requesters.paginators.pagination_strategy import PaginationStrategy
+from airbyte_cdk.sources.declarative.requesters.request_option import RequestOption, RequestOptionType
+from airbyte_cdk.sources.declarative.requesters.request_options.interpolated_request_options_provider import (
+    InterpolatedRequestOptionsProvider,
+)
+from airbyte_cdk.sources.declarative.types import Config
+
+
+class LimitPaginator(ConditionalPaginator):
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/limit_paginator.py]: you're right. Moving the stop_condition to the strategy allows us to get rid of the conditional paginator

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/cursor_pagination_strategy.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        :param decoder: decoder to decode the response
+        """
+        if isinstance(cursor_value, str):
+            cursor_value = InterpolatedString(cursor_value)
+        self._cursor_value = cursor_value
+        self._config = config
+        self._decoder = decoder or JsonDecoder()
+
+    def next_page_token(self, response: requests.Response, last_records: List[Mapping[str, Any]]) -> Optional[Any]:
+        token = self._cursor_value.eval(config=self._config, last_records=last_records, decoded_response=self._decoder.decode(response))
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/cursor_pagination_strategy.py:49]: good idea. will do in [this PR](https://github.com/airbytehq/airbyte/pull/14877) after we merge this in

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/page_increment.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, List, Mapping, Optional
+
+import requests
+from airbyte_cdk.sources.declarative.requesters.paginators.pagination_strategy import PaginationStrategy
+
+
+class PageIncrement(PaginationStrategy):
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/page_increment.py:11]: updated as per https://github.com/airbytehq/airbyte/pull/14433#discussion_r925493698

**Review by sherifnada — APPROVED** [review]
looks great!

**Pull request merged**